File: Farmsy/Core/TripEndpoints.swift
import Foundation
import CoreLocation

/// Where a saved drive starts and ends (R1), and the one place that shape is
/// enforced on its way to and from `trips`.
///
/// The bug this closes: a trip is a name and a list of farms, so `save()` never
/// carried the ends. `setOrigin` writes the start to UserDefaults under
/// `dlb_trip_origin` and no further — `TripRow` is `(user_id, name)`. Save a
/// drive from Utrecht to Groningen, reopen it tomorrow, and it starts at the
/// first farm instead of at your house. Nothing errors; the trip just quietly
/// means something else than what was saved.
///
/// The web has the same rules in `src/lib/tripEndpoints.ts`, and migration 058
/// has them again as CHECK constraints. Three implementations of one truth, the
/// same shape as the access rule in T-1 — if a rule changes here it changes
/// there, in the same commit.
///
/// Three rules that are easy to get wrong:
///  - **A coordinate is a pair or it is nothing.** Half a point survives a write
///    and then fails at the far end of a route calculation, where the cause is
///    no longer visible.
///  - **Null is legal, everywhere.** Every trip saved before R1 has no ends, and
///    a trip that is only a list of farms stays a valid trip. Reading one back
///    must produce a usable plan, not an error.
///  - **The radius travels with the pair.** A corridor only means something at
///    the width it was chosen at: five kilometres around a city drive and five
///    around a cross-country one are different offers. Reopening at the default
///    width would quietly show a different set of farms than the one saved.

/// Somewhere someone chose in the place search. Kept separate from
/// `CLLocationCoordinate2D` so it can carry the label it was chosen by, and so
/// this file is testable without a map.
struct TripPlace: Equatable, Codable {
    let lat: Double
    let lng: Double
    /// The place name as it was picked. Stored rather than re-geocoded: Photon
    /// and MKLocalSearch will not always give back the same words for the same
    /// point, and a saved trip that renames its own start is unsettling.
    let label: String

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }

    /// A pair that is really on the planet, or nil. Rejects the half-pair, NaN
    /// and infinity — all three survive a write and fail somewhere else.
    static func make(lat: Double?, lng: Double?, label: String?) -> TripPlace? {
        guard let lat, let lng, lat.isFinite, lng.isFinite else { return nil }
        guard (-90...90).contains(lat), (-180...180).contains(lng) else { return nil }

        // A point with no label is still a point. Falling back to the
        // coordinates is uglier than a place name and better than dropping a
        // start someone chose.
        let text = label?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return TripPlace(
            lat: lat,
            lng: lng,
            label: text.isEmpty ? String(format: "%.4f, %.4f", lat, lng) : text
        )
    }

    static func make(_ coordinate: CLLocationCoordinate2D, label: String?) -> TripPlace? {
        make(lat: coordinate.latitude, lng: coordinate.longitude, label: label)
    }
}

struct TripEndpoints: Equatable {
    var origin: TripPlace?
    var destination: TripPlace?
    /// Corridor width in km. Nil means "never chosen" — the panel suggests one.
    var radiusKm: Double?
    /// The day it was planned for, `yyyy-mm-dd`. Nil for anything saved before
    /// R7, and for a trip that is only a list of farms.
    ///
    /// A date and not a timestamp. A farm shop's opening hours are a property of
    /// the weekday, not of an instant, and a `Date` would drag a time zone into
    /// a question that does not have one. Saturday in Amsterdam is Saturday.
    var date: String?
    /// When the drive sets off, minutes past midnight. Nil means never chosen.
    var departMinutes: Int?

    static let none = TripEndpoints(
        origin: nil, destination: nil, radiusKm: nil, date: nil, departMinutes: nil
    )

    /// The widest corridor the clients offer. Anything past it is a bad write
    /// rather than a preference, so it is dropped instead of clamped: clamping
    /// stores a number nobody chose.
    static let maxRadiusKm: Double = 100

    static func validRadius(_ km: Double?) -> Double? {
        guard let km, km.isFinite, km > 0, km <= maxRadiusKm else { return nil }
        return km
    }

    /// A `yyyy-mm-dd` that is actually one, or nil.
    ///
    /// Shape-checked and then round-tripped through a calendar, because
    /// `2026-02-31` matches the pattern and is not a day. A trip that claims to
    /// be planned for a date that does not exist would compare against whatever
    /// weekday the calendar rolled it into, which is a wrong answer delivered
    /// with confidence.
    static func validDate(_ text: String?) -> String? {
        guard let text, text.count == 10 else { return nil }
        let parts = text.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 3,
              parts[0].count == 4, parts[1].count == 2, parts[2].count == 2,
              let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2]),
              text.allSatisfy({ $0.isNumber || $0 == "-" })
        else { return nil }

        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "Europe/Amsterdam") ?? .current
        guard let made = cal.date(from: DateComponents(year: y, month: m, day: d)) else { return nil }
        let back = cal.dateComponents([.year, .month, .day], from: made)
        return back.year == y && back.month == m && back.day == d ? text : nil
    }

    /// Minutes past midnight, whole, inside a day. Matches the 059 CHECK, so a
    /// bad value is dropped here rather than rejected by Postgres after a
    /// round trip.
    static func validDepartMinutes(_ minutes: Int?) -> Int? {
        guard let minutes, (0...1439).contains(minutes) else { return nil }
        return minutes
    }

    /// The calendar every trip date is read and written in.
    ///
    /// Fixed to Amsterdam rather than the device's. A trip is planned against
    /// Dutch and Belgian opening hours, and a phone set to Los Angeles would
    /// otherwise turn a Saturday drive into a Friday one at the date boundary —
    /// the farm would be reported shut and nobody would know why.
    static var calendar: Calendar {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "Europe/Amsterdam") ?? .current
        return cal
    }

    /// A `Date` as the `yyyy-mm-dd` the column stores.
    static func dayString(_ date: Date) -> String {
        let c = calendar.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    }

    /// A stored `yyyy-mm-dd` back as a `Date` at local noon, or nil.
    ///
    /// Noon and not midnight: a date pinned to midnight lands on the wrong day
    /// the moment anything shifts it by an hour, and a daylight-saving boundary
    /// does exactly that twice a year.
    static func day(from text: String?) -> Date? {
        guard let text = validDate(text) else { return nil }
        let parts = text.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3 else { return nil }
        return calendar.date(from: DateComponents(
            year: parts[0], month: parts[1], day: parts[2], hour: 12
        ))
    }

    /// The day the planner offers when nobody has chosen one: the Saturday
    /// coming, today included if today is Saturday. Mirrors `nextSaturday()` in
    /// the web's MapSearchContext — a farm trip is a weekend errand, and
    /// defaulting to a weekday would answer a question nobody asked.
    static func nextSaturday(from: Date = Date()) -> String {
        let cal = calendar
        // Calendar counts Sunday as 1; Saturday is 7.
        let weekday = cal.component(.weekday, from: from)
        let ahead = (7 - weekday) % 7
        let day = cal.date(byAdding: .day, value: ahead, to: from) ?? from
        return dayString(day)
    }
}

/// The `trips` columns from migrations 058 and 059, exactly as they are stored.
///
/// Encodable for the insert, Decodable for reading a saved trip back. All
/// optional: a trip from before R1 decodes to `TripEndpoints.none`, which is a
/// working plan and not a failure.
struct TripEndpointRow: Codable, Equatable {
    var origin_lat: Double?
    var origin_lng: Double?
    var origin_label: String?
    var destination_lat: Double?
    var destination_lng: Double?
    var destination_label: String?
    var radius_km: Double?
    var trip_date: String?
    var depart_minutes: Int?

    /// Always all nine, so clearing an end writes NULL rather than leaving
    /// yesterday's value behind. The failure mode of a partial update is a trip
    /// that keeps a start the user deliberately removed.
    init(_ endpoints: TripEndpoints) {
        origin_lat        = endpoints.origin?.lat
        origin_lng        = endpoints.origin?.lng
        origin_label      = endpoints.origin?.label
        destination_lat   = endpoints.destination?.lat
        destination_lng   = endpoints.destination?.lng
        destination_label = endpoints.destination?.label
        radius_km         = TripEndpoints.validRadius(endpoints.radiusKm)
        trip_date         = TripEndpoints.validDate(endpoints.date)
        depart_minutes    = TripEndpoints.validDepartMinutes(endpoints.departMinutes)
    }

    var endpoints: TripEndpoints {
        TripEndpoints(
            origin: TripPlace.make(lat: origin_lat, lng: origin_lng, label: origin_label),
            destination: TripPlace.make(lat: destination_lat, lng: destination_lng, label: destination_label),
            radiusKm: TripEndpoints.validRadius(radius_km),
            // Postgres hands a `date` back as `yyyy-mm-dd`, which is what the
            // planner holds, so there is nothing to convert — only to check.
            date: TripEndpoints.validDate(trip_date),
            departMinutes: TripEndpoints.validDepartMinutes(depart_minutes)
        )
    }

    /// The columns to select when opening a trip. Here so a new column cannot be
    /// added to the row without the query that reads it.
    static let columns =
        "origin_lat, origin_lng, origin_label, destination_lat, destination_lng, destination_label, radius_km, trip_date, depart_minutes"
}

File: android/app/src/main/java/app/farmsy/android/core/TripEndpoints.kt
package app.farmsy.android.core

import com.google.android.gms.maps.model.LatLng
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.format.ResolverStyle
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObjectBuilder
import kotlinx.serialization.json.JsonPrimitive

/// Where a saved drive starts and ends (R1), and the one place that shape is
/// enforced on its way to and from `trips`.
///
/// The bug this closes: a trip is a name and a list of farms, so save() never
/// carried the ends. setOrigin() writes the start to SharedPreferences and no
/// further. Save a drive from Utrecht to Groningen, reopen it tomorrow, and it
/// starts at the first farm instead of at your house. Nothing errors; the trip
/// just quietly means something else than what was saved.
///
/// The same rules live in `Farmsy/Core/TripEndpoints.swift`, in
/// `src/lib/tripEndpoints.ts` on the web, and again as CHECK constraints in
/// migration 058. Four places, one truth — the same shape as the T-1 access
/// matrix. If a rule changes here it changes there, in the same commit.
///
/// Three rules that are easy to get wrong:
///  - **A coordinate is a pair or it is nothing.** Half a point survives a write
///    and then fails at the far end of a route calculation, where the cause is
///    no longer visible.
///  - **Null is legal, everywhere.** Every trip saved before R1 has no ends, and
///    a trip that is only a list of farms stays a valid trip. Reading one back
///    must produce a usable plan, not an error.
///  - **The radius travels with the pair.** A corridor only means something at
///    the width it was chosen at: five kilometres around a city drive and five
///    around a cross-country one are different offers. Reopening at the default
///    width would quietly show a different set of farms than the one saved.

/// Somewhere someone chose in the place search. Separate from [LatLng] so it can
/// carry the label it was picked by — stored rather than re-geocoded, because
/// Photon will not always give back the same words for the same point and a
/// saved trip that renames its own start is unsettling.
data class TripPlace(
    val lat: Double,
    val lng: Double,
    val label: String,
) {
    val latLng: LatLng get() = LatLng(lat, lng)

    companion object {
        /// A pair that is really on the planet, or null. Rejects the half-pair,
        /// NaN and infinity — all three survive a write and fail elsewhere.
        fun make(lat: Double?, lng: Double?, label: String?): TripPlace? {
            if (lat == null || lng == null) return null
            if (!lat.isFinite() || !lng.isFinite()) return null
            if (lat < -90.0 || lat > 90.0 || lng < -180.0 || lng > 180.0) return null

            // A point with no label is still a point. Falling back to the
            // coordinates is uglier than a place name and better than dropping a
            // start someone chose.
            val text = label?.trim().orEmpty()
            return TripPlace(
                lat = lat,
                lng = lng,
                label = if (text.isEmpty()) String.format("%.4f, %.4f", lat, lng) else text,
            )
        }

        fun make(coord: LatLng?, label: String?): TripPlace? =
            if (coord == null) null else make(coord.latitude, coord.longitude, label)
    }
}

data class TripEndpoints(
    val origin: TripPlace? = null,
    val destination: TripPlace? = null,
    /// Corridor width in km. Null means "never chosen" — the panel suggests one.
    val radiusKm: Double? = null,
    /// The day it was planned for, `yyyy-mm-dd`. Null for anything saved before
    /// R7, and for a trip that is only a list of farms.
    ///
    /// This is what makes "drive this again" able to say what changed. Without
    /// a day to compare against there is nothing to compare: every farm would
    /// report today, and the answer would be a saved route against nothing.
    val date: String? = null,
    /// When it set off, minutes past midnight. Null means never chosen.
    val departMinutes: Int? = null,
) {
    companion object {
        val NONE = TripEndpoints()

        /// The widest corridor the clients offer. Anything past it is a bad
        /// write rather than a preference, so it is dropped instead of clamped:
        /// clamping stores a number nobody chose.
        const val MAX_RADIUS_KM = 100.0

        fun validRadius(km: Double?): Double? =
            if (km != null && km.isFinite() && km > 0.0 && km <= MAX_RADIUS_KM) km else null

        /// A trip is planned against Dutch and Belgian opening hours, so the day
        /// is decided in Amsterdam and not on whatever clock the phone is set
        /// to. A device in Los Angeles would otherwise turn a Saturday drive
        /// into a Friday one at the date boundary.
        val zone: ZoneId = ZoneId.of("Europe/Amsterdam")

        /// `yyyy-mm-dd`, or null. Validated against a calendar and not only a
        /// pattern: `2026-02-31` matches the shape and is not a day. A lenient
        /// parser rolls it into March, and the trip then compares opening hours
        /// against the wrong weekday and reports it with full confidence.
        fun validDate(text: String?): String? {
            val trimmed = text?.trim().orEmpty()
            if (trimmed.length != 10) return null
            return try {
                LocalDate.parse(trimmed, ISO_DAY).toString().takeIf { it == trimmed }
            } catch (_: DateTimeParseException) {
                null
            }
        }

        /// Minutes past midnight inside one day, or null. The bounds are the
        /// same ones migration 059 states as a CHECK — 1440 is tomorrow, not a
        /// time today.
        fun validDepartMinutes(minutes: Int?): Int? =
            if (minutes != null && minutes in 0..1439) minutes else null

        /// A stored day as a [LocalDate], or null if it is not a day.
        fun day(text: String?): LocalDate? =
            validDate(text)?.let { LocalDate.parse(it, ISO_DAY) }

        fun dayString(date: LocalDate): String = date.format(ISO_DAY)

        /// The Saturday coming, today included when today is Saturday.
        ///
        /// Offered by the picker and never stored on its own: a trip saved
        /// before R7 has no day, and inventing one would make it claim it was
        /// planned for a date nobody chose. Mirrors `nextSaturday()` in the
        /// web's MapSearchContext.
        fun nextSaturday(from: LocalDate = LocalDate.now(zone)): String {
            val ahead = (DayOfWeek.SATURDAY.value - from.dayOfWeek.value + 7) % 7
            return dayString(from.plusDays(ahead.toLong()))
        }

        private val ISO_DAY: DateTimeFormatter =
            DateTimeFormatter.ofPattern("uuuu-MM-dd").withResolverStyle(ResolverStyle.STRICT)
    }
}

/// The `trips` columns from migration 058, exactly as they are stored.
///
/// All nullable: a trip from before R1 decodes to [TripEndpoints.NONE], which is
/// a working plan and not a failure.
@Serializable
data class TripEndpointRow(
    @SerialName("origin_lat") val originLat: Double? = null,
    @SerialName("origin_lng") val originLng: Double? = null,
    @SerialName("origin_label") val originLabel: String? = null,
    @SerialName("destination_lat") val destinationLat: Double? = null,
    @SerialName("destination_lng") val destinationLng: Double? = null,
    @SerialName("destination_label") val destinationLabel: String? = null,
    @SerialName("radius_km") val radiusKm: Double? = null,
    @SerialName("trip_date") val tripDate: String? = null,
    @SerialName("depart_minutes") val departMinutes: Int? = null,
) {
    val endpoints: TripEndpoints
        get() = TripEndpoints(
            origin = TripPlace.make(originLat, originLng, originLabel),
            destination = TripPlace.make(destinationLat, destinationLng, destinationLabel),
            radiusKm = TripEndpoints.validRadius(radiusKm),
            date = TripEndpoints.validDate(tripDate),
            departMinutes = TripEndpoints.validDepartMinutes(departMinutes),
        )

    companion object {
        /// Always all nine, so clearing an end writes NULL rather than leaving
        /// yesterday's value behind. The failure mode of a partial update is a
        /// trip that keeps a start the user deliberately removed.
        ///
        /// A day or a departure that does not survive validation is dropped
        /// here rather than sent. The CHECK in migration 059 would reject the
        /// whole row, so one bad minute would cost the entire save: a trip with
        /// no day still saves, a trip that fails to save is gone.
        fun from(endpoints: TripEndpoints) = TripEndpointRow(
            originLat = endpoints.origin?.lat,
            originLng = endpoints.origin?.lng,
            originLabel = endpoints.origin?.label,
            destinationLat = endpoints.destination?.lat,
            destinationLng = endpoints.destination?.lng,
            destinationLabel = endpoints.destination?.label,
            radiusKm = TripEndpoints.validRadius(endpoints.radiusKm),
            tripDate = TripEndpoints.validDate(endpoints.date),
            departMinutes = TripEndpoints.validDepartMinutes(endpoints.departMinutes),
        )

        /// The columns to select when opening a trip. Here so a new column cannot
        /// be added to the row without the query that reads it.
        const val COLUMNS =
            "origin_lat, origin_lng, origin_label, destination_lat, destination_lng, destination_label, radius_km, trip_date, depart_minutes"
    }
}

/// The nine end columns, written onto a `trips` insert or update.
///
/// Always all nine, including the nulls. A partial update would leave
/// yesterday's start on a trip whose start the user deliberately removed, and
/// that is a worse bug than the one R1 fixes: it is wrong rather than missing.
fun JsonObjectBuilder.putEndpoints(row: TripEndpointRow) {
    fun num(v: Double?) = if (v == null) JsonNull else JsonPrimitive(v)
    fun int(v: Int?) = if (v == null) JsonNull else JsonPrimitive(v)
    fun str(v: String?) = if (v == null) JsonNull else JsonPrimitive(v)

    put("origin_lat", num(row.originLat))
    put("origin_lng", num(row.originLng))
    put("origin_label", str(row.originLabel))
    put("destination_lat", num(row.destinationLat))
    put("destination_lng", num(row.destinationLng))
    put("destination_label", str(row.destinationLabel))
    put("radius_km", num(row.radiusKm))
    put("trip_date", str(row.tripDate))
    put("depart_minutes", int(row.departMinutes))
}
